src/util.rs

```diff
@@ -139,6 +139,65 @@ pub(crate) mod ptr {
             // - `U: 'a`
             Ptr { ptr, _lifetime: PhantomData }
         }
+
+        /// Project a field from `self`.
+        ///
+        /// # Safety
+        ///
+        /// ## Preconditions
+        ///
+        /// The caller promises that `projector` projects a well-aligned field
+        /// of its argument. Its argument will be `self` casted to a raw
+        /// pointer.
+        ///
+        /// ## Postconditions
+        ///
+        /// If the preconditions of this function are met, this function will
+        /// return a `Ptr` the field projected from `self` by `projector`.
+        #[doc(hidden)]
+        #[inline]
+        pub unsafe fn project<U: 'a + ?Sized>(
+            self,
+            projector: impl FnOnce(*mut T) -> *mut U,
+        ) -> Ptr<'a, U> {
+            // SAFETY: `projector` is provided with `self` casted to a raw
+            // pointer.
+            let field = projector(self.ptr.as_ptr());
+
+            // SAFETY: The caller of `project` promises that `projector` is
+            // provided with `self` casted to a raw pointer. By invariant,
+            // `self` is non-null. By contract, `projector` is a field
+            // projection of its argument. The field projected by `projector`
+            // will, too, be non-null.
```

Comment:
There's a bit in here that is off or insufficient IMO:
- The caller doesn't promise that `projector` is provided with a raw pointer of `self`, we do
- We should explicitly call out that field projection can't wrap around the address space to null. You can cite the field invariant on `ptr` and could also leave a TODO to cite [these docs](https://github.com/rust-lang/rust/pull/116675) once they're merged.

Reply:
Done.